Fix for issue #431 - Covariance is not transformed in do_transform_pose_with_covariance_stamped

oss-vineet opened the pull request:
-Fixed the issue of covariance not being transformed in do_transform_pose_with_covariance_stamped, corresponding to #431 and made the same changes in Python as was done in #430 

-Added the tests for do_transform_pose_with_covariance_stamped

oss-vineet committed:
-Fixed the issue of covariance not being transformed in do_transform_pose_with_covariance_stamped, corresponding to https://github.com/ros2/geometry2/issues/431 and made the same changes as the C++ API
-Added the tests for do_transform_pose_with_covariance_stamped

aprotyas commented on tf2_geometry_msgs/src/tf2_geometry_msgs/tf2_geometry_msgs.py:
```diff
@@ -62,6 +63,89 @@ def transform_to_kdl(t):
+    cov_22 = np.array([cov_in[21:24], cov_in[27:30], cov_in[33:]])
+
+    # And we perform the transform
+    result_11 = R * cov_11 * R_transpose
+    result_12 = R * cov_12 * R_transpose
+    result_21 = R * cov_21 * R_transpose
+    result_22 = R * cov_22 * R_transpose
```
For `np.array` types which represent 2D matrices, the `*` operator performs element-wise multiplication, and not the conventional matrix product that we're expecting here. We can get the desired multiplication by using `@` instead.

```suggestion
    result_11 = R @ cov_11 @ R_transpose
    result_12 = R @ cov_12 @ R_transpose
    result_21 = R @ cov_21 @ R_transpose
    result_22 = R @ cov_22 @ R_transpose
```

oss-vineet replied:
Yes this was my mistake. Thanks for correcting, I will resolve this

aprotyas commented on tf2_geometry_msgs/scripts/test.py:
```diff
@@ -135,6 +135,42 @@ def test_transform(self):
+        v.pose.pose.orientation = Quaternion(w=0.0, x=1.0, y=0.0, z=0.0)
+
+        out = tf2_geometry_msgs.do_transform_pose_with_covariance_stamped(v, t)
+        expected_covariance = (
+          1.0, 0.0, 0.0, 4.0, 0.0, 0.0,
+          0.0, 2.0, 0.0, 0.0, 5.0, 0.0,
+          0.0, 0.0, 3.0, 0.0, 0.0, 6.0,
+          1.0, 0.0, 0.0, 4.0, 0.0, 0.0,
+          0.0, 2.0, 0.0, 0.0, 5.0, 0.0,
+          0.0, 0.0, 3.0, 0.0, 0.0, 6.0
+        )
```
Can you double check the math here? I did some matrix multiplications by hand, and I think this should instead read:
```suggestion
        expected_covariance = (
          1.0, -2.0, -3.0, 4.0, -5.0, -6.0,
          -1.0, 2.0, 3.0, -4.0, 5.0, 6.0,
          -1.0, 2.0, 3.0, -4.0, 5.0, 6.0,
          1.0, -2.0, -3.0, 4.0, -5.0, -6.0,
          -1.0, 2.0, 3.0, -4.0, 5.0, 6.0,
          -1.0, 2.0, 3.0, -4.0, 5.0, 6.0
        )
```

oss-vineet replied:
Yeah I did the math by hand now and indeed you are right. Just for this part, I tried calculating it on the python terminal and due to my mistake with multiplying np arrays, I got this wrong, my bad.

aprotyas commented on tf2_geometry_msgs/scripts/test.py:
```diff
@@ -135,6 +135,42 @@ def test_transform(self):
+          0.0, 2.0, 0.0, 0.0, 5.0, 0.0,
+          0.0, 0.0, 3.0, 0.0, 0.0, 6.0
+        )
+        self.assertEqual(out.pose.pose.position.x, 2)
+        self.assertEqual(out.pose.pose.position.y, 0)
+        self.assertEqual(out.pose.pose.position.z, 0)
+        self.assertEqual(out.pose.covariance, expected_covariance)
```
You should also check that the output orientations are correct.
```suggestion
        self.assertEqual(out.pose.pose.position.x, 2)
        self.assertEqual(out.pose.pose.position.y, 0)
        self.assertEqual(out.pose.pose.position.z, 0)
        self.assertEqual(out.pose.orientation.x, 0)
        self.assertEqual(out.pose.orientation.y, 0)
        self.assertEqual(out.pose.orientation.z, 0)
        self.assertEqual(out.pose.orientation.w, 1)
        self.assertEqual(out.pose.covariance, expected_covariance)
```

aprotyas replied:
For line 173 specifically: `out.pose.covariance` is an `np.array` with shape `(36,)`, while `expected_covariance` is a tuple. As such, their comparison won't really work. The comparison here will return an `np.array` with shape `(36,)` filled with boolean values corresponding to element-wise comparisons - so a quick hack could be to 
```py
        self.assertTrue(np.all(out.pose.covariance == expected_covariance))
```
Another solution could be to convert `expected_covariance` to a `numpy` type, and then use [`np.array_equiv`](https://numpy.org/doc/stable/reference/generated/numpy.array_equiv.html#numpy.array_equiv).

oss-vineet replied:
When I try the above in terminal, it gives me the following:
```
>>> np.all(out.pose.covariance == expected_covariance)
<stdin>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
False
```
So I will change the `expected_covariance` to a `np.array` of shape `(36,)`

aprotyas reviewed (changes requested):
Requesting some changes for `numpy` eccentricities and math reviews.

aprotyas commented on tf2_geometry_msgs/src/tf2_geometry_msgs/tf2_geometry_msgs.py:
```diff
@@ -62,6 +63,89 @@ def transform_to_kdl(t):
+    q1 = transform.transform.rotation.x
+    q2 = transform.transform.rotation.y
+    q3 = transform.transform.rotation.z
+     
+    # First row of the rotation matrix
+    r00 = 2 * (q0 * q0 + q1 * q1) - 1
+    r01 = 2 * (q1 * q2 - q0 * q3)
+    r02 = 2 * (q1 * q3 + q0 * q2)
+     
+    # Second row of the rotation matrix
+    r10 = 2 * (q1 * q2 + q0 * q3)
+    r11 = 2 * (q0 * q0 + q2 * q2) - 1
+    r12 = 2 * (q2 * q3 - q0 * q1)
+     
+    # Third row of the rotation matrix
+    r20 = 2 * (q1 * q3 - q0 * q2)
+    r21 = 2 * (q2 * q3 + q0 * q1)
+    r22 = 2 * (q0 * q0 + q3 * q3) - 1
+
+    R =  np.array([[r00, r01, r02],
+                   [r10, r11, r12],
+                   [r20, r21, r22]])
+    
```
There is trailing whitespace in the empty lines. Can you remove them please?

aprotyas replied:
2e960db

Commit:
Update tf2_geometry_msgs/scripts/test.py

Correct the expected test results for covariance matrix

Co-authored-by: Abrar Rahman Protyasha <aprotyas@u.rochester.edu>

Commit:
Update tf2_geometry_msgs/src/tf2_geometry_msgs/tf2_geometry_msgs.py

Correct np array multiplication operator

Co-authored-by: Abrar Rahman Protyasha <aprotyas@u.rochester.edu>

oss-vineet committed:
Committing the changes requested in PR #453: https://github.com/ros2/geometry2/pull/453
- Remove whitespaces between lines
- Add a few more helpful comments
- Check assertion on pose.orientation as well
- Correct check for array equivalence and make expected_covariance as np.array

aprotyas commented on tf2_geometry_msgs/scripts/test.py:
```diff
@@ -159,18 +159,22 @@ def test_transform(self):
+        self.assertEqual(out.pose.orientation.y, 0)
+        self.assertEqual(out.pose.orientation.z, 0)
+        self.assertEqual(out.pose.orientation.w, 1)
+        self.assertTrue(np.array_equiv(out.pose.covariance, expected_covariance))
```
In hindsight, the purpose of the comparison here may fit [`np.array_equal`](https://numpy.org/doc/stable/reference/generated/numpy.array_equal.html) (check for shape and element equality) over [`np.array_equiv`](https://numpy.org/doc/stable/reference/generated/numpy.array_equiv.html) (check for shape broadcastable and element equality) - since we want to avoid the output pose covariance from being ill-formed during the transformation step, i.e. we should check that the `do_transform` function does not return an output covariance that is not in the `(X,)` shape.

Thoughts?

oss-vineet replied:
Also in hindsight, I was just wondering whether I should've mentioned it in my previous comment, but I thought since you made the suggestion, you would've been aware of it. But reading the documentation of both, it definitely seems like `np.array_equal` is the choice to go by, as I cannot think of a situation where we would need to broadcast some other array shape into this shape since the covariance will always be `(36,)`. So I will make this change.

And also something I thought of mentioning earlier is that the `@` operator for matrix multiplication is only available from Python 3.5 onwards (which has already been EOL'd) and the alternative for numpy matrix multiplication is `A.dot(B)`.

aprotyas replied:
> is only available from Python 3.5 onwards

I believe that's fine because of what you said, and also because the default Python version in current OS's tends to be higher than that.

aprotyas reviewed (commented):
Small concern about array-equality operations, check comments inline please.

oss-vineet committed:
As per the PR discussions for #453, just a small change from np.array_equiv to np.array_equal since we should also check if the shapes of the array are equal.

aprotyas reviewed (approved):
LGTM 👍

clalancette commented on tf2_geometry_msgs/src/tf2_geometry_msgs/tf2_geometry_msgs.py:
```diff
@@ -62,6 +63,95 @@ def transform_to_kdl(t):
                                     t.transform.translation.y,
                                     t.transform.translation.z))
 
+def transformCovariance(cov_in, transform):
```
Nit: the style in this file is snake_case, so I'd prefer this to be `transform_covariance`.

oss-vineet replied:
Right, done!

clalancette reviewed (changes requested):
One minor change, then I think this will be ready for CI.

oss-vineet committed:
Minor style change

clalancette commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15314)](http://ci.ros2.org/job/ci_linux/15314/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10026)](http://ci.ros2.org/job/ci_linux-aarch64/10026/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12994)](http://ci.ros2.org/job/ci_osx/12994/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15507)](http://ci.ros2.org/job/ci_windows/15507/)

clalancette commented:
All right, CI is looking good.  While the crashes on macOS are concerning, they also happened in the nightlies: https://ci.ros2.org/view/nightly/job/nightly_osx_release/2207/#showFailuresLink .  So I'm going to go ahead and merge this, thanks for the contribution.
